File: RB110/ruby_small_problems/Easy7/swap_case.rb
=begin
Write a method that takes a string as an argument and returns a new string
in which every uppercase letter is replaced by its lowercase version, and
every lowercase letter by its uppercase version. All other characters should
be unchanged.

You may not use String#swapcase; write your own version of this method.

EXAMPLE:
swapcase('PascalCase') == 'pASCALcASE'
swapcase('Tonight on XYZ-TV') == 'tONIGHT ON xyz-tv'

------------------------------
--- P
- GIVEN: string argument
- RETURN: NEW string
          - where every uppcased letter is downcased
          - every downcased letter is upcased
          - every other char is left unchanged
          - cannot use swapcase
  - what happens if string is empty?
  - what happens if non string arg is given?
  - 
--- E
swapcase('Tonight on XYZ-TV') == 'tONIGHT ON xyz-tv'
  - Uppcased chars = 'TXYZTV' turns into 'txyztv'
  - lowcased chars = 'onighton' turns into 'ONIGHTON'
  - unchanged chars = '  -' 
    - whitespace and any nonalphabetical chars such as special/numeric left unchanged


--- D
-beginning: string
intermediate: array would allow transofrmation over each char
ending: string
--- A
- CONVERT string into array of 'words'
  - split into individual chars
- TRANSFORM chars based on previously defined rules
  - for each char, #map
  - if the current char is equal to the char lowercased
    - uppercase the char #upcase
  - else
    - lowercase the char #downcase

- CONVERT array back into string
  - join(' ')
- RETURN string

=end

def swapcase(string)
  string.chars.map do |char|
    (char == char.downcase) ? char.upcase : char.downcase
  end.join
end

p swapcase('PascalCase') == 'pASCALcASE'
p swapcase('Tonight on XYZ-TV') == 'tONIGHT ON xyz-tv'

File: LS215/JS_Small_Problems/string_and_text_processing/swap_case.js
// Write a function that takes a string as an argument and returns that string with every lowercase letter changed to uppercase and every uppercase letter changed to lowercase. Leave all other characters unchanged.

// Examples:

function swapCase(string) {
  let newString = '';

  for (let idx = 0; idx < string.length; idx += 1) {
    if (/[A-Z]/.test(string[idx])) newString += string[idx].toLowerCase();
    if (/[a-z]/.test(string[idx])) newString += string[idx].toUpperCase();
    if (/[^a-z]/i.test(string[idx])) newString += string[idx];
  }

  return newString;
}

console.log(swapCase('CamelCase'));              // "cAMELcASE"
console.log(swapCase('Tonight on XYZ-TV'));      // "tONIGHT ON xyz-tv"